### DISCUSSION TITLE
Use for custom dataset

### DISCUSSION THREAD

**Discussion by MZeinab** [roboflow/supervision#21] [Q&A]
### Search before asking

- [X] I have searched the Supervision [issues](https://github.com/roboflow/supervision/issues) and found no similar bug report.

### Bug

Thanks, it was so amazing. 
I have one question. I'll be so grateful if you could help:
1) I watched the video that explained how to train yolov5 with the custom dataset. I used that notebook and exported the best.pt file, containing model weights. My question is: how to use it for inference?
- I  used model = torch.hub.load('ultralytics/yolov5', 'custom', path='best.pt')  # PyTorch
but It loads the original .pt file, not my custom one.

### Environment

colab

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by github-actions** [comment]
Hello there, thank you for opening an Issue ! 🙏🏻 The team was notified and they will get back to you asap.

**Comment by SkalskiP** [comment]
Hello there as it is not really a problem with Supervision, I'll convert that issue into discussion.

**Comment by SkalskiP** [comment]
@MZeinab did you try to pass a global path to the weights file?

**Reply by MZeinab** [reply]
what do you mean by global path?
I use google colab. I uploaded my custom weight file "best.pt" and used it in 
torch.hub.load('ultralytics/yolov5', 'custom', path='best.pt') .
Anything more should I do?
I think the problem is with 'ultralytics/yolov5'.
Do you have any idea? I'll be so thankful if you could help.

**Reply by SkalskiP** [reply]
Hi, @MZeinab 👋🏻!

> I think the problem is with `ultralytics/yolov5`.

It is possible, but before we create an issue on their repository, let's try to pass the global path. Let me try to explain. Let's try to imagine this directory structure (I just made it up to give you an example):

```
content/
├─ yolov5/
├─ weights/
│  ├─ best.pt
```

Now if you are inside the `weights` directory path to your `best.pt` can be expressed in the form of a global and relative paths. 

- global `/content/weights/best.pt`
- relative `best.pt`

Problem is that when you are inside `yolov5` directory relative path is no longer `best.pt` but `../best.pt`. That's why it is always better to use global paths. They are longer but stay the same regardless of your active directory.

I can help you to resolve your global path to `best.pt` file if you upload a screenshot of your directory structure here. ;)

**Comment by MZeinab** [comment]
Hi @SkalskiP 
I just wanted to let you know that I checked what you explained. I used the global path.

- [ ] This is my best.pt file:

`file = "/content/best.pt"
model_real = torch.load(file)
model_real`
 The output is:

{'epoch': -1, 'best_fitness': None, 'model': Model(
   (model): Sequential(
     (0): Conv(
       (conv): Conv2d(3, 32, kernel_size=(6, 6), stride=(2, 2), padding=(2, 2), bias=False)
       (bn): BatchNorm2d(32, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
       (act): SiLU(inplace=True)
     )
     (1): Conv(
       (conv): Conv2d(32, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
       (bn): BatchNorm2d(64, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
       (act): SiLU(inplace=True)
     )
........... (I didn't add the remain of it)



- [ ] I compared it with the code I mentioned earlier :

`model = torch.hub.load('ultralytics/yolov5', 'custom', path="/content/best.pt")
model`
the output is:

Using cache found in /root/.cache/torch/hub/ultralytics_yolov5_master
YOLOv5 🚀 2023-2-18 Python-3.8.10 torch-1.13.1+cu116 CUDA:0 (Tesla T4, 15110MiB)

Fusing layers... 
Model summary: 213 layers, 7012822 parameters, 0 gradients, 15.8 GFLOPs
Adding AutoShape... 
AutoShape(
  (model): DetectMultiBackend(
    (model): DetectionModel(
      (model): Sequential(
        (0): Conv(
          (conv): Conv2d(3, 32, kernel_size=(6, 6), stride=(2, 2), padding=(2, 2))
          (act): SiLU(inplace=True)
        )
        (1): Conv(
          (conv): Conv2d(32, 64, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1))
          (act): SiLU(inplace=True)
        )



you can see that the batch normalization layers are not in the custom model (model_real) and It seems that the code uses the original model of ultralytics

**Reply by SkalskiP** [reply]
`torch.hub.load` might load just selected layers. Did you try to run the actual inference?

**Reply by MZeinab** [reply]
yeah, I think so.
How can I use the best.pt file to convert it to model.
I used 
`file = "/content/best.pt"

model_real = torch.load(file)`

and runned this part:

`import supervision as sv

### extract video frame
generator = sv.get_video_frames_generator('/content/Multi_Ablated_25a25c_Exp21_6day.mp4')
iterator = iter(generator)
frame = next(iterator)

### detect
results = model_real(frame, size=1280)
detections = sv.Detections.from_yolov5(results)

### annotate
box_annotator = sv.BoxAnnotator(thickness=4, text_thickness=4, text_scale=2)
frame = box_annotator.annotate(scene=frame, detections=detections)

%matplotlib inline  
sv.show_frame_in_notebook(frame, (16, 16))`

####
but It gets error:
'dict' object is not callable

so my question is how to define model_real based on my best.pt file to be callable?

**Reply by SkalskiP** [reply]
@MZeinab, but did you actually try to infer using `torch.hub.load` or just loaded?

If you want to load YOLOv5 model not via `torch.hub.load` I think your best choice is to use YOLOv5 [pip package](https://pypi.org/project/yolov5/).

**Reply by MZeinab** [reply]
yes, I tried to infer the model defined with `torch.hub.load`. 
but the detection result was not good,
It detects some bounding boxes that are not the desired object.

so after that when I compared this 2 .pt file (my custom model and the yolov5 acheived with `torch.hub.load` ) I found out that they are different. 
and after that I defined my custom model with `model_real = torch.load(file)` and try to infer but I get error that 'dict' object is not callable
